File: assets/mathlets/heat_1d.py
import numpy as np

# =====================================================
# Heat equation on [0, L] with Dirichlet boundary values
#   u_t = k u_xx,   u(0,t)=U0,  u(L,t)=UL
# u = u_s + v, with u_s(x) = U0 + (UL-U0) x/L
# v has homogeneous Dirichlet BC and sine-series solution.
# =====================================================

L = 10.0
NX = 800
N_COEFF = 200
N_SHOW = 5

X = np.linspace(0.0, L, NX)

# fixed y-scale
Y_MIN, Y_MAX = -5.5, 5.5

MODE_COLORS = ["#d2f8d2", "#9ee9a1", "#66d36e", "#2fb94a", "#128c2b"]

def steady_state(X, U0, UL):
    return U0 + (UL - U0) * (X / L)

def u0_uniform(X, C):
    return C * np.ones_like(X)

def u0_gaussian(X, A, x0, sigma):
    return A * np.exp(-0.5 * ((X - x0) / sigma) ** 2)

def u0_linear(X, U0, UL):
    return steady_state(X, U0, UL)

def fourier_coeffs(gx, nmax):
    # S shape: (nmax, NX)
    S = np.sin(np.outer(np.arange(1, nmax + 1), np.pi * X / L))
    bn = (2.0 / L) * np.trapz(gx * S, X, axis=1)
    return bn

def solution_profile(t, k, U0, UL, ic_kind, C, A, x0, sigma, n_coeff=N_COEFF):
    if ic_kind == "uniform":
        u0 = u0_uniform(X, C)
    elif ic_kind == "gaussian":
        u0 = u0_gaussian(X, A, x0, sigma)
    elif ic_kind == "linear":
        u0 = u0_linear(X, U0, UL)
    else:
        u0 = u0_uniform(X, 0.0)

    us = steady_state(X, U0, UL)
    gx = u0 - us
    bn = fourier_coeffs(gx, n_coeff)

    n = np.arange(1, n_coeff + 1)[:, None]              # (n_coeff, 1)
    lam = (np.pi * n / L) ** 2                          # (n_coeff, 1)
    decay = np.exp(-k * lam * t)                        # (n_coeff, 1)
    S = np.sin(n * np.pi * X[None, :] / L)              # (n_coeff, NX)
    v = (bn[:, None] * decay * S).sum(axis=0)           # (NX,)

    return (us + v), us, bn

def compute_plot_data(ic_kind, show_modes, t, U0, UL, kappa, C, A, x0, sigma):
    """
    Returns data for Plotly:
      X, u, us, modes, L, Y_MIN, Y_MAX

    modes is a list of [y_mode, color, name] for n=1..5 if enabled.
    """
    t = float(t)
    U0 = float(U0); UL = float(UL)
    kappa = float(kappa)
    C = float(C); A = float(A)
    x0 = float(x0); sigma = float(sigma)

    u, us, bn = solution_profile(t, kappa, U0, UL, ic_kind, C, A, x0, sigma, n_coeff=N_COEFF)

    modes_out = []
    if int(show_modes) == 1:
        n = np.arange(1, N_SHOW + 1)
        lam = (np.pi * n / L) ** 2
        decay = np.exp(-kappa * lam * t)
        for i in range(N_SHOW):
            y_mode = us + bn[i] * decay[i] * np.sin((i + 1) * np.pi * X / L)
            modes_out.append([y_mode.tolist(), MODE_COLORS[i], f"mode n={i+1}"])

    return X.tolist(), u.tolist(), us.tolist(), modes_out, float(L), float(Y_MIN), float(Y_MAX)

File: projects/edo-edp-mathlets/Heat_1D/heat_1d.js
let py = null;

// same green gradient as Dash code
const MODE_COLORS = ["#d2f8d2", "#9ee9a1", "#66d36e", "#2fb94a", "#128c2b"];

function fmt(x, d = 2) { return Number(x).toFixed(d); }

// error helper
function showErr(e) {
  const box = document.getElementById("errBox");
  if (!box) return;
  box.style.display = "block";
  box.textContent = (e && e.stack) ? e.stack : String(e);
}
function clearErr() {
  const box = document.getElementById("errBox");
  if (!box) return;
  box.style.display = "none";
  box.textContent = "";
}

async function initPyodideAndModule() {
  py = await initPyodideBase({
    packages: ["numpy"],
    stderr: (s) => console.log("[pyodide]", s)
  });

  // expects:
  // ../../../assets/mathlets/heat-1d.py
  await loadPythonFile(py, "../../../assets/mathlets/heat_1d.py");
}

function updateDashSliderUI(input) {
  const min = Number(input.min);
  const max = Number(input.max);
  const val = Number(input.value);

  const pct = ((val - min) / (max - min)) * 100;

  const wrap = input.closest(".slider-wrap");
  if (!wrap) return;

  wrap.style.setProperty("--fill", String(pct));

  const ticks = wrap.querySelectorAll(".dash-tick");
  ticks.forEach(t => {
    const pStr = t.style.getPropertyValue("--p").trim();
    const p = Number(pStr.replace("%", ""));
    t.classList.toggle("active", p <= pct + 1e-9);
  });
}

function toggleICBlocks(kind) {
  const uni = document.getElementById("ic-uniform-params");
  const gauss = document.getElementById("ic-gaussian-params");
  const note = document.getElementById("ic-linear-note");

  if (kind === "uniform") {
    uni.style.display = "block";
    gauss.style.display = "none";
    note.style.display = "none";
  } else if (kind === "gaussian") {
    uni.style.display = "none";
    gauss.style.display = "block";
    note.style.display = "none";
  } else {
    uni.style.display = "none";
    gauss.style.display = "none";
    note.style.display = "block";
  }
}

async function computeProfile(params) {
  // params: {ic_kind, show_modes, t, U0, UL, kappa, C, A, x0, sigma}
  py.globals.set("IC", params.ic_kind);
  py.globals.set("SHOW", params.show_modes ? 1 : 0);

  py.globals.set("t", params.t);
  py.globals.set("U0", params.U0);
  py.globals.set("UL", params.UL);
  py.globals.set("kappa", params.kappa);

  py.globals.set("C", params.C);
  py.globals.set("A", params.A);
  py.globals.set("x0", params.x0);
  py.globals.set("sigma", params.sigma);

  const out = py.runPython(`
compute_plot_data(IC, SHOW, t, U0, UL, kappa, C, A, x0, sigma)
  `);

  return out.toJs();
  // returns:
  // [X, u, us, modes, L, Y_MIN, Y_MAX]
  // modes = list of [y_mode, color, name]
}

async function redraw() {
  try {
    clearErr();

    const ic_kind = document.getElementById("ic-kind").value;
    const show_modes = document.getElementById("show-modes").checked;

    const t = Number(document.getElementById("t").value);
    const U0 = Number(document.getElementById("U0").value);
    const UL = Number(document.getElementById("UL").value);
    const kappa = Number(document.getElementById("kappa").value);

    const C = Number(document.getElementById("C").value);
    const A = Number(document.getElementById("A").value);
    const x0 = Number(document.getElementById("x0").value);
    const sigma = Number(document.getElementById("sigma").value);

    // readouts
    document.getElementById("tVal").textContent = fmt(t, 1);
    document.getElementById("U0Val").textContent = fmt(U0, 2);
    document.getElementById("ULVal").textContent = fmt(UL, 2);
    document.getElementById("kVal").textContent = fmt(kappa, 2);
    document.getElementById("CVal").textContent = fmt(C, 2);
    document.getElementById("AVal").textContent = fmt(A, 2);
    document.getElementById("x0Val").textContent = fmt(x0, 2);
    document.getElementById("sigmaVal").textContent = fmt(sigma, 2);

    const [X, u, us, modes, L, Y_MIN, Y_MAX] = await computeProfile({
      ic_kind, show_modes, t, U0, UL, kappa, C, A, x0, sigma
    });

    const traces = [
      {
        type: "scatter",
        mode: "lines",
        x: X, y: u,
        line: { width: 2.6 },
        name: `u(x,t=${t.toFixed(1)})`
      },
      {
        type: "scatter",
        mode: "lines",
        x: X, y: us,
        line: { width: 2.0, dash: "dash" },
        name: "steady state"
      }
    ];

    if (modes && modes.length) {
      for (const m of modes) {
        const y_mode = m[0];
        const color = m[1];
        const name = m[2];
        traces.push({
          type: "scatter",
          mode: "lines",
          x: X, y: y_mode,
          line: { width: 1.5, color },
          name
        });
      }
    }

    const layout = {
      template: "plotly_dark",
      margin: { l: 55, r: 10, t: 10, b: 55 },
      xaxis: { title: "x", range: [0, L], showgrid: true, gridcolor: "#444", zeroline: false, autorange: false },
      yaxis: { title: "u(x,t)", range: [Y_MIN, Y_MAX], showgrid: true, gridcolor: "#444", zeroline: false, autorange: false },
      paper_bgcolor: "#1e1e1e",
      plot_bgcolor: "#1e1e1e",
      font: { color: "#f0f0f0" },
      legend: { bgcolor: "rgba(0,0,0,0.35)" }
    };

    await Plotly.react("plot", traces, layout, { responsive: true });

  } catch (e) {
    showErr(e);
    console.error(e);
  }
}

function setupLanguageToggle() {
  const langBtn = document.getElementById("langToggle");
  const textES = document.getElementById("text-es");
  const textEN = document.getElementById("text-en");
  if (!langBtn || !textES || !textEN) return;

  let currentLang = "es";
  langBtn.addEventListener("click", () => {
    if (currentLang === "es") {
      textES.style.display = "none";
      textEN.style.display = "block";
      langBtn.textContent = "Español";
      currentLang = "en";
    } else {
      textEN.style.display = "none";
      textES.style.display = "block";
      langBtn.textContent = "English";
      currentLang = "es";
    }
    if (window.MathJax) MathJax.typesetPromise();
  });
}

async function main() {
  setupLanguageToggle();

  await initPyodideAndModule();

  // init slider UI state
  const sliderIds = ["t", "U0", "UL", "kappa", "C", "A", "x0", "sigma"];
  sliderIds.forEach(id => updateDashSliderUI(document.getElementById(id)));

  // listeners
  sliderIds.forEach(id => {
    const el = document.getElementById(id);
    el.addEventListener("input", () => {
      updateDashSliderUI(el);
      redraw();
    });
  });

  const ic = document.getElementById("ic-kind");
  ic.addEventListener("change", () => {
    toggleICBlocks(ic.value);
    redraw();
  });

  const show = document.getElementById("show-modes");
  show.addEventListener("change", () => redraw());

  // initial block visibility
  toggleICBlocks(ic.value);

  await redraw();
}

main();